watchdog: Add watchdog driver

CristXu opened the pull request:
Add watchdog driver for the RT1050/60

Signed-off-by: Crist Xu <crist.xu@nxp.com>

Fixes #27122

CristXu committed:
watchdog: Add watchdog driver

Add watchdog driver for the RT1050/60

Signed-off-by: Crist Xu <crist.xu@nxp.com>

MaureenHelm commented:
The buildkite check is failing because the hal_nxp module needs to be updated:
```diff
diff --git a/mcux/drivers/imx/CMakeLists.txt b/mcux/drivers/imx/CMakeLists.txt
index 2cac2a4..66b741d 100644
--- a/mcux/drivers/imx/CMakeLists.txt
+++ b/mcux/drivers/imx/CMakeLists.txt
@@ -29,3 +29,4 @@ zephyr_library_sources_ifdef(CONFIG_PWM_MCUX          fsl_pwm.c)
 zephyr_library_sources_ifdef(CONFIG_SPI_MCUX_LPSPI     fsl_lpspi.c)
 zephyr_library_sources_ifdef(CONFIG_UART_MCUX_LPUART   fsl_lpuart.c)
 zephyr_library_sources_ifdef(CONFIG_VIDEO_MCUX_CSI     fsl_csi.c)
+zephyr_library_sources_ifdef(CONFIG_WDT_MCUX_IMX_WDOG  fsl_wdog.c)
```

Please see [changes to existing modules](https://docs.zephyrproject.org/latest/guides/modules.html#changes-to-existing-modules) for more details how to do this.

MaureenHelm commented on drivers/watchdog/wdt_mcux_imx_wdog.c:
```diff
@@ -0,0 +1,167 @@
+	WDOG_GetDefaultConfig(&data->wdog_config);
+
+	data->wdog_config.timeoutValue =
+		  WDOG_TMOUT_SEC(cfg->window.max / 1000.0f);
```
Does the divisor need to be a float? Can it be an integer instead?

CristXu replied:
NO, I think float is better, becasue our watchdog's resolution is 0.5(timeout periods from 0.5 to 128), but the register can only accept (0 - 127), so  i write a macro:#define WDOG_TMOUT_SEC(x)  ((int)(x * 2) - 1) that can accept a float. if the divisor is a integer, maybe we can not accept a float timeout value, such as 0.5(if we use integer divisor, the result maybe 500ms / 1000 = 0, has no effect), maybe a wasting of the resolution.

MaureenHelm replied:
If you multiply by two before the divide, you can achieve the same result in fixed point.

CristXu replied:
Yes，You are right， I will rewrite the macro, and update this after

CristXu replied:
Has updated，Please check

MaureenHelm commented on west.yml:
```diff
@@ -98,7 +98,7 @@ manifest:
       path: tools/net-tools
     - name: hal_nxp
-      revision: 7276046215cfcd4ed3d6723f16f7aeb868b0fb60
+      revision: pull/51/head
```
The nxp_hal PR has now been merged, so you can update this to the new SHA

CristXu replied:
ok， i will update this

pabigot commented on drivers/watchdog/wdt_mcux_imx_wdog.c:
```diff
@@ -0,0 +1,167 @@
+/*
+ *
+ * Copyright (c) 2018, NXP
```
Correct year and comment style:
```
/*
 * Copyright (C) 2020, NXP
 *
 * SPDX-License-Identifier: Apache-2.0
 */
```

pabigot commented on drivers/watchdog/wdt_mcux_imx_wdog.c:
```diff
@@ -0,0 +1,167 @@
+#include <logging/log.h>
+LOG_MODULE_REGISTER(wdt_mcux_wdog);
+
+#define WDOG_TMOUT_SEC(x)  (x * 2 / 1000 - 1)
```
I think this is taking the configured window size in milliseconds and returning the floor of half the duration, in seconds.

Using `MSEC_PER_SEC` instead of 1000 would be better if that's correct.

CristXu replied:
> I think this is taking the configured window size in milliseconds and returning the floor of half the duration, in seconds.
> 
> Using `MSEC_PER_SEC` instead of 1000 would be better if that's correct.

Hi，forget an (int) to cast the value. Has Updated with (int) and also the MSEC_PER_SEC

pabigot reviewed (dismissed):
A couple minor code concerns.  Also `wdt_basic_test` doesn't get past the first phase.

pabigot commented on drivers/watchdog/wdt_mcux_imx_wdog.c:
```diff
@@ -0,0 +1,166 @@
+#include <logging/log.h>
+LOG_MODULE_REGISTER(wdt_mcux_wdog);
+
+#define WDOG_TMOUT_SEC(x)  ((int)(x * 2 / MSEC_PER_SEC) - 1)
```
The cast to `int` is not necessary.

pabigot replied:
The cleanup is ok, but the test still doesn't pass on mimxrt1050_evk

CristXu replied:
Updated with no cast to int

MaureenHelm commented:
> Also `wdt_basic_test` doesn't get past the first phase.

@CristXu will you please investigate why the test isn't passing anymore?

CristXu commented:
> > Also `wdt_basic_test` doesn't get past the first phase.
> 
> @CristXu will you please investigate why the test isn't passing anymore?

Hi, I saw that all the auto-check  have passed, Could you please tell me what does the test you mean? I do not know where to check this result.

pabigot commented:
`tests/drivers/watchdog/wdt_basic_api` on frdm_k64f gives this output:

```
Running test suite wdt_basic_test
===================================================================
START - test_wdt
Testcase: test_wdt_no_callback
Waiting to restart MCU
Running test suite wdt_basic_test
===================================================================
START - test_wdt
Testcase: test_wdt_no_callback
Testcase passed
Testcase: test_wdt_callback_1
Waiting to restart MCU
Running test suite wdt_basic_test
===================================================================
START - test_wdt
Testcase: test_wdt_callback_1
Testcase passed
Testcase: test_wdt_bad_window_max
E: Invalid timeout
 PASS - test_wdt
===================================================================
Test suite wdt_basic_test succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

When I run the same test on this branch on `mimxrt1060_evk` I get:

```
Running test suite wdt_basic_test
===================================================================
START - test_wdt
Testcase: test_wdt_no_callback
Waiting to restart MCU
Running test suite wdt_basic_test
===================================================================
START - test_wdt
Testcase: test_wdt_no_callback
Waiting to restart MCU
Running test suite wdt_basic_test
===================================================================
START - test_wdt
Testcase: test_wdt_no_callback
Waiting to restart MCU
Running test suite wdt_basic_test
===================================================================
START - test_wdt
Testcase: test_wdt_no_callback
Waiting to restart MCU
... // repeats
```
which means the test is not reaching the callback test and never prints the PASS message.

CristXu commented:
> `tests/drivers/watchdog/wdt_basic_api` on `frdm_k64f` gives this output:
> 
> ```
> Running test suite wdt_basic_test
> ===================================================================
> START - test_wdt
> Testcase: test_wdt_no_callback
> Waiting to restart MCU
> Running test suite wdt_basic_test
> ===================================================================
> START - test_wdt
> Testcase: test_wdt_no_callback
> Testcase passed
> Testcase: test_wdt_callback_1
> Waiting to restart MCU
> Running test suite wdt_basic_test
> ===================================================================
> START - test_wdt
> Testcase: test_wdt_callback_1
> Testcase passed
> Testcase: test_wdt_bad_window_max
> E: Invalid timeout
>  PASS - test_wdt
> ===================================================================
> Test suite wdt_basic_test succeeded
> ===================================================================
> PROJECT EXECUTION SUCCESSFUL
> ```
> 
> When I run the same test on this branch on `mimxrt1060_evk` I get:
> 
> ```
> Running test suite wdt_basic_test
> ===================================================================
> START - test_wdt
> Testcase: test_wdt_no_callback
> Waiting to restart MCU
> Running test suite wdt_basic_test
> ===================================================================
> START - test_wdt
> Testcase: test_wdt_no_callback
> Waiting to restart MCU
> Running test suite wdt_basic_test
> ===================================================================
> START - test_wdt
> Testcase: test_wdt_no_callback
> Waiting to restart MCU
> Running test suite wdt_basic_test
> ===================================================================
> START - test_wdt
> Testcase: test_wdt_no_callback
> Waiting to restart MCU
> ... // repeats
> ```
> 
> which means the test is not reaching the callback test and never prints the PASS message.

Ok, I will check this.

CristXu commented:
> `tests/drivers/watchdog/wdt_basic_api` on `frdm_k64f` gives this output:
> 
> ```
> Running test suite wdt_basic_test
> ===================================================================
> START - test_wdt
> Testcase: test_wdt_no_callback
> Waiting to restart MCU
> Running test suite wdt_basic_test
> ===================================================================
> START - test_wdt
> Testcase: test_wdt_no_callback
> Testcase passed
> Testcase: test_wdt_callback_1
> Waiting to restart MCU
> Running test suite wdt_basic_test
> ===================================================================
> START - test_wdt
> Testcase: test_wdt_callback_1
> Testcase passed
> Testcase: test_wdt_bad_window_max
> E: Invalid timeout
>  PASS - test_wdt
> ===================================================================
> Test suite wdt_basic_test succeeded
> ===================================================================
> PROJECT EXECUTION SUCCESSFUL
> ```
> 
> When I run the same test on this branch on `mimxrt1060_evk` I get:
> 
> ```
> Running test suite wdt_basic_test
> ===================================================================
> START - test_wdt
> Testcase: test_wdt_no_callback
> Waiting to restart MCU
> Running test suite wdt_basic_test
> ===================================================================
> START - test_wdt
> Testcase: test_wdt_no_callback
> Waiting to restart MCU
> Running test suite wdt_basic_test
> ===================================================================
> START - test_wdt
> Testcase: test_wdt_no_callback
> Waiting to restart MCU
> Running test suite wdt_basic_test
> ===================================================================
> START - test_wdt
> Testcase: test_wdt_no_callback
> Waiting to restart MCU
> ... // repeats
> ```
> 
> which means the test is not reaching the callback test and never prints the PASS message.
Hi， I have checked the code, and found that when place data on ocram / sdram, the test will fail, but the code under samples/drivers/watchdog can run(the second img); but when i place the data on dtcm, the test can pass(the first img). I think that the watchdog trigger a systemreset which will effect the time sequence of the memory controller(
Affect data read and write), and there is a value **volatile uint32_t m_state __attribute__((section(".noinit.test_wdt")));**  used to switch the callback function stored in this memory, so after the system reset, the value of this one maybe unpredictible，especially on sdram. As a result, the code will always run the first **Testcase: test_wdt_no_callback**.
![image](https://user-images.githubusercontent.com/49855018/92368083-d227d380-f129-11ea-93f4-64b9e4bca4ea.png)
![image](https://user-images.githubusercontent.com/49855018/92368123-e1a71c80-f129-11ea-82de-29997e5bc8e9.png)

pabigot commented:
> Hi， I have checked the code, and found that when place data on ocram / sdram, the test will fail, but the code under samples/drivers/watchdog can run(the second img); but when i place the data on dtcm, the test can pass(the first img). I think that the watchdog trigger a systemreset which will effect the time sequence of the memory controller(
> Affect data read and write), and there is a value **volatile uint32_t m_state **attribute**((section(".noinit.test_wdt")));**  used to switch the callback function stored in this memory, so after the system reset, the value of this one maybe unpredictible，especially on sdram. As a result, the code will always run the first **Testcase: test_wdt_no_callback**.

That sounds reasonable.  Then the default configuration for this test should place the data in memory where it's retained over the watchdog.  I don't even know how to move the data section; is that documented somewhere?

But the test needs to pass, or an issue will be immediately opened to fix it.

CristXu commented:
> > Hi， I have checked the code, and found that when place data on ocram / sdram, the test will fail, but the code under samples/drivers/watchdog can run(the second img); but when i place the data on dtcm, the test can pass(the first img). I think that the watchdog trigger a systemreset which will effect the time sequence of the memory controller(
> > Affect data read and write), and there is a value **volatile uint32_t m_state **attribute**((section(".noinit.test_wdt")));**  used to switch the callback function stored in this memory, so after the system reset, the value of this one maybe unpredictible，especially on sdram. As a result, the code will always run the first **Testcase: test_wdt_no_callback**.
> 
> That sounds reasonable. Then the default configuration for this test should place the data in memory where it's retained over the watchdog. I don't even know how to move the data section; is that documented somewhere?
> 
> But the test needs to pass, or an issue will be immediately opened to fix it.

Yes, the key_word **attribute((section(".noinit.test_wdt")));** used to place the varible to the **.noinit_test_wdt** section, this section is placed in sram(defined by DATA_LOCATION ), So maybe we can set a section on DTCM by __attribute__((section(".dtcm.test_wdt"))), and also alloctate this on the linker file like this.
![image](https://user-images.githubusercontent.com/49855018/92398632-8c83fe80-f15b-11ea-8356-b97958066163.png)
Or you can look this: 
https://stackoverflow.com/questions/11180892/force-gcc-to-forgo-zeroing-certain-globals/11181026#11181026

pabigot commented:
I think https://github.com/pabigot/zephyr/commit/23e27095a547094ed0b524b6a366b18d16d95d7d would be an acceptable modification to the watchdog test.

However, the mimxrt1060_evk board doesn't set the `zephyr,dtcm` chosen that's the standard way of identifying presence of this feature across platforms.  If it's added I get linker errors about multiply defined sections.

In the process of testing this I seem to have bricked my MIMXRT1060-EVK: in `reset.S` upon executing the fifth or sixth instruction it jumps to 0x002020b4 which doesn't have anything.  Neither `blinky` nor `hello_world` work anymore, removing power doesn't help.

So I don't know what's going on, but unless I can get my board working again I can't help with this anymore.

CristXu commented:
> I seem to have bricked my MIMXRT1060-EVK

Maybe forget the definition of this section(.dtcm_noinit.test_wdt) in the link file? iff not, this new section will overlap the original area.

CristXu commented:
> I think [pabigot@23e2709](https://github.com/pabigot/zephyr/commit/23e27095a547094ed0b524b6a366b18d16d95d7d) would be an acceptable modification to the watchdog test.
> 
> However, the mimxrt1060_evk board doesn't set the `zephyr,dtcm` chosen that's the standard way of identifying presence of this feature across platforms. If it's added I get linker errors about multiply defined sections.
> 
> In the process of testing this I seem to have bricked my MIMXRT1060-EVK: in `reset.S` upon executing the fifth or sixth instruction it jumps to 0x002020b4 which doesn't have anything. Neither `blinky` nor `hello_world` work anymore, removing power doesn't help.
> 
> So I don't know what's going on, but unless I can get my board working again I can't help with this anymore.

Hi，About this Bug，i have a new discovery. It seems that the data can not be placed in a cached-area, such as ocram or sdram, because before the system reset, we do not clean the cache to update the memory, but the watchdog will reset the cache-controller maybe, so after reset, the dirty data will loss. This is also why the code works when we put the data to the DTCM, the dtcm is not a cache area. So, I think there are two ways to fix: 1. place the data  on a non-cache area, e.g. dtcm, 2. clean the cache   befor the reset, it's better to put the function in the watchdog handler, but when we do not set the cb func, this will fail.

pabigot commented:
There is already standard `.dtcm_noinit.*` section support in the Zephyr linker scripts for persisting state across resets when the device supports DTCM.  At least some iMX boards don't define the `zephyr,dtcm` chosen that's required to enable it.  I haven't tried to figure out how DTCM gets used on those boards today.

I still think the right fix is https://github.com/pabigot/zephyr/commit/23e27095a547094ed0b524b6a366b18d16d95d7d because it'd be the best chance for a cross-platform solution.

CristXu commented:
> There is already standard `.dtcm_noinit.*` section support in the Zephyr linker scripts for persisting state across resets when the device supports DTCM. At least some iMX boards don't define the `zephyr,dtcm` chosen that's required to enable it. I haven't tried to figure out how DTCM gets used on those boards today.
> 
> I still think the right fix is [pabigot@23e2709](https://github.com/pabigot/zephyr/commit/23e27095a547094ed0b524b6a366b18d16d95d7d) because it'd be the best chance for a cross-platform solution.

hi， agree with you

dleach02 commented:
anyone have a clue why buildkite is failing? I tried building this branch locally and it built fine.

pabigot commented on dts/arm/nxp/nxp_rt.dtsi:
```diff
@@ -670,6 +670,13 @@
 			status = "disabled";
 			#address-cells = <1>;
 			#size-cells = <0>;
+
```
@dleach02 a closing `};` got removed here, causing a paren balance error breaking the parse.

CristXu replied:
Yes, Missing a }; here, I delete it carelessly when i rebase my branch, i will update this quickly.

dleach02 replied:
now I'm confused why it doesn't fail for me locally???

dleach02 replied:
Wait, nvm, I see what I did... I'm stupid with git and didn't actually checkout his branch. Sigh.

dleach02 reviewed (approved):
LGTM

pabigot reviewed (commented):
I don't see the test changes that I thought were necessary, and since I don't want to brick my 1060 again I'm not going to try to run the test.  So I've dropped my NACK, but I'm not going to approve/disapprove.

dleach02 commented:
> I don't see the test changes that I thought were necessary, and since I don't want to brick my 1060 again I'm not going to try to run the test. So I've dropped my NACK, but I'm not going to approve/disapprove.

@pabigot, while I appreciate the concern (I also had a bricked board), the problem was an inter rim

MaureenHelm commented:
> > I don't see the test changes that I thought were necessary, and since I don't want to brick my 1060 again I'm not going to try to run the test. So I've dropped my NACK, but I'm not going to approve/disapprove.
> 
> @pabigot, while I appreciate the concern (I also had a bricked board), the problem was with a broken code from a while ago in the history. As long as that code is not run you will be okay.
> 
> @CristXu, can you respond to Peter's concern about the test changes?

The PR as-is doesn't pass the test on `mimxrt1050_evk`. (It doesn't brick the board either)

CristXu commented:
> > I don't see the test changes that I thought were necessary, and since I don't want to brick my 1060 again I'm not going to try to run the test. So I've dropped my NACK, but I'm not going to approve/disapprove.
> 
> @pabigot, while I appreciate the concern (I also had a bricked board), the problem was with a broken code from a while ago in the history. As long as that code is not run you will be okay.
> 
> @CristXu, can you respond to Peter's concern about the test changes?

Hi，Maureen，Peter，Sorry for the late update, i always think that peter will update this PR, sorry again. I have updated the code according peter's method, Thank you peter. and it works, meanwhile, my board is lucky that it does not brick. so i think the code can run now.
Here is the result, but user must add the zephyr,dtcm=&dtcm inside the chosen. like this: 
![image](https://user-images.githubusercontent.com/49855018/96665993-c6fecf00-1388-11eb-83e9-41a6a2b2c4b5.png)
I delete this line when i push the code
![image](https://user-images.githubusercontent.com/49855018/96665908-8dc65f00-1388-11eb-942f-355b3cb60b3b.png)

pabigot reviewed (changes requested):
Yes, it was trying to get dtcm enabled on mimxrt1060-evk that caused the brickage.

I rebased #29385 on master, then this on that, and the test passes, but it does so with an error log message about an invalid timeout.  That should be investigated.

```Running test suite wdt_basic_test
===================================================================
START - test_wdt
Testcase: test_wdt_no_callback
Waiting to restart MCU
Running test suite wdt_basic_test
===================================================================
START - test_wdt
Testcase: test_wdt_no_callback
Testcase passed
Testcase: test_wdt_callback_1
Waiting to restart MCU
Running test suite wdt_basic_test
===================================================================
START - test_wdt                                                                
Testcase: test_wdt_callback_1                                                   
Testcase passed                                                                 
Testcase: test_wdt_bad_window_max                                               
E: Invalid timeoutValue, valid (0.5s - 128.0s)                                  
 PASS - test_wdt                                                                
===================================================================             
Test suite wdt_basic_test succeeded                                             
===================================================================             
PROJECT EXECUTION SUCCESSFUL   
```

CristXu commented:
> Yes, it was trying to get dtcm enabled on mimxrt1060-evk that caused the brickage.
> 
> I rebased #29385 on master, then this on that, and the test passes, but it does so with an error log message about an invalid timeout. That should be investigated.
> 
> ```
> ===================================================================
> START - test_wdt
> Testcase: test_wdt_no_callback
> Waiting to restart MCU
> Running test suite wdt_basic_test
> ===================================================================
> START - test_wdt
> Testcase: test_wdt_no_callback
> Testcase passed
> Testcase: test_wdt_callback_1
> Waiting to restart MCU
> Running test suite wdt_basic_test
> ===================================================================
> START - test_wdt                                                                
> Testcase: test_wdt_callback_1                                                   
> Testcase passed                                                                 
> Testcase: test_wdt_bad_window_max                                               
> E: Invalid timeoutValue, valid (0.5s - 128.0s)                                  
>  PASS - test_wdt                                                                
> ===================================================================             
> Test suite wdt_basic_test succeeded                                             
> ===================================================================             
> PROJECT EXECUTION SUCCESSFUL   
> ```

Hi, because the third test will pass  and set: 	
        m_cfg_wdt0.window.max = 0U;
	err = wdt_install_timeout(wdt, &m_cfg_wdt0);
That is really a bad parameter. So i think the code is right. and it also show a PASS - test_wdt

pabigot commented:
> Hi, because the third test will pass and set:
> m_cfg_wdt0.window.max = 0U;
> err = wdt_install_timeout(wdt, &m_cfg_wdt0);
> That is really a bad parameter. So i think the code is right. and it also show a PASS - test_wdt
> And the log is from the watchdog driver.

OK, I understand now.  Thanks.

galak reviewed (commented):
What about wdog1?

looking at some of the user manuals, seems like there is more than one instance, any reason not to add `wdog1` into dts?

galak commented on dts/bindings/watchdog/nxp,imx-wdog.yaml:
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2018, NXP
+# SPDX-License-Identifier: Apache-2.0
+
+description: imxRT watchdog
```
nit: looks like the same wdog on i.mx6/7, so probably good to just say `i.MX watchdog (including i.MX RT1xxx, but not include i.MX- RT6/6xx)`

CristXu commented:
> What about wdog1?
> 
> looking at some of the user manuals, seems like there is more than one instance, any reason not to add `wdog1` into dts?

HI, yes, in fact, there are 3 watchdog in RT, watchdog1, watchdog2, and rtwatchdog(wd3), but from the RM, you can saw that there are 2 diferent IP(watchdog1+2 and rtwatchdog), and about the watchdog1 and 2, the RM said **Upon timeout, the WDOG1 asserts the
internal system reset signal, WDOG_RESET_B_DEB to the System Reset Controller
(SRC); and WDOG2 asserts interrupt to SNVS, to report the security violation condition.** so i choose the watchdog1.
and the rtwatchdog is from Kinetics in fact, so it is diferent with the others.

galak commented:
> HI, yes, in fact, there are 3 watchdog in RT, watchdog1, watchdog2, and rtwatchdog(wd3), but from the RM, you can saw that there are 2 diferent IP(watchdog1+2 and rtwatchdog), and about the watchdog1 and 2, the RM said **Upon timeout, the WDOG1 asserts the internal system reset signal, WDOG_RESET_B_DEB to the System Reset Controller (SRC); and WDOG2 asserts interrupt to SNVS, to report the security violation condition.** so i choose the watchdog1.
> and the rtwatchdog is from Kinetics in fact, and the wd1/wd2 is from i.mx, so all the 3 wd were different, and also with diferent SDK function.

is there any reason not to add watchdog2 to the devicetree and have it disabled?  I get the 3rd watchdog is different and would need a different driver (rtwatchdog).

CristXu commented:
> WDOG2 asserts interrupt to SNVS, to report the security violation condition

Hi, The WDOG2 will assert an interrupt to SNVS, not the same as the WDOG1. The RM said **WDOG1 will be configured during boot while WDOG2 is dedicated for secure world
purposes and will be activated by TZ software if required.** So i leave the WODG2, only set the WDOG1, in case that user will use this wdog(WDOG2).

CristXu commented:
> > HI, yes, in fact, there are 3 watchdog in RT, watchdog1, watchdog2, and rtwatchdog(wd3), but from the RM, you can saw that there are 2 diferent IP(watchdog1+2 and rtwatchdog), and about the watchdog1 and 2, the RM said **Upon timeout, the WDOG1 asserts the internal system reset signal, WDOG_RESET_B_DEB to the System Reset Controller (SRC); and WDOG2 asserts interrupt to SNVS, to report the security violation condition.** so i choose the watchdog1.
> > and the rtwatchdog is from Kinetics in fact, and the wd1/wd2 is from i.mx, so all the 3 wd were different, and also with diferent SDK function.
> 
> is there any reason not to add watchdog2 to the devicetree and have it disabled? I get the 3rd watchdog is different and would need a different driver (rtwatchdog).

Hi, The WDOG2 will assert an interrupt to SNVS, not the same as the WDOG1. The RM said **WDOG1 will be configured during boot while WDOG2 is dedicated for secure world
purposes and will be activated by TZ software if required.** So i leave the WODG2, only set the WDOG1, in case that user will use this wdog(WDOG2).

galak commented:
> Hi, The WDOG2 will assert an interrupt to SNVS, not the same as the WDOG1. The RM said **WDOG1 will be configured during boot while WDOG2 is dedicated for secure world purposes and will be activated by TZ software if required.** So i leave the WODG2, only set the WDOG1, in case that user will use this wdog(WDOG2).

I understand, but what TZ software are we talking about?  There is no TZ on the Cortex-M7s. (feels like that statement is left over from i.MX 6/7/8)  I still think it makes sense to have the devicetree represented the hardware of the SoC.  WDOG2 would be `disabled` so it shouldn't impact anything.

CristXu commented:
> > Hi, The WDOG2 will assert an interrupt to SNVS, not the same as the WDOG1. The RM said **WDOG1 will be configured during boot while WDOG2 is dedicated for secure world purposes and will be activated by TZ software if required.** So i leave the WODG2, only set the WDOG1, in case that user will use this wdog(WDOG2).
> 
> I understand, but what TZ software are we talking about? There is no TZ on the Cortex-M7s. (feels like that statement is left over from i.MX 6/7/8) I still think it makes sense to have the devicetree represented the hardware of the SoC. WDOG2 would be `disabled` so it shouldn't impact anything.

Hi，Yes, no TZ on the Cortex-M7s, so the TZ is not the same thing. but i do not know exactly what it really is.  And agree with you, i will add the other WDOG node.

CristXu commented:
> > > Hi, The WDOG2 will assert an interrupt to SNVS, not the same as the WDOG1. The RM said **WDOG1 will be configured during boot while WDOG2 is dedicated for secure world purposes and will be activated by TZ software if required.** So i leave the WODG2, only set the WDOG1, in case that user will use this wdog(WDOG2).
> > 
> > 
> > I understand, but what TZ software are we talking about? There is no TZ on the Cortex-M7s. (feels like that statement is left over from i.MX 6/7/8) I still think it makes sense to have the devicetree represented the hardware of the SoC. WDOG2 would be `disabled` so it shouldn't impact anything.
> 
> Hi，Yes, no TZ on the Cortex-M7s, so the TZ is not the same thing. but i do not know exactly what it really is. And agree with you, i will add the other WDOG node.

Hi, Updated, Please check.
